### DISCUSSION TITLE
In app purchase controls / library for mac app store / windows app store?

### DISCUSSION THREAD

**Discussion by lanyusan** [AvaloniaUI/Avalonia#9636] [Q&A]
Hello,

Is there some in app purchase  controls / library for mac app store / windows app store for Avalonia?

There is an In-App Billing Plugin for .NET MAUI, Xamarin, and Windows:

https://github.com/jamesmontemagno/InAppBillingPlugin

Unfortunately it doesn't support avalonia. 

Does anyone know if there is anything similar for Avalonia out there, or could anyone share some insights/blogs etc. about implementing iap with Avalonia?

**Comment by robloo** [comment]
Here is my general plan for Windows: https://techcommunity.microsoft.com/t5/windows-dev-appconsult/enable-in-app-product-purchases-for-desktop-bridge-converted/ba-p/316901. Avalonia is essentially a classic .net or Win32 app in this case. Microsoft does allow desktop bridge apps to access these API's. It's going to take some work though and there is no drop-in library I've found yet.

macOS is TBD but someone else may know more. There must be a way to do this with native code interop though.

This type of thing is ideal for a Toolkit. Or adding Avalonia support to other projects like:  https://github.com/jamesmontemagno/InAppBillingPlugin which you already found (and is recommended by Microsoft here https://github.com/dotnet/maui/issues/1461)

**Reply by lanyusan** [reply]
Had a nice conversion through an issue with James Montemagno.

https://github.com/jamesmontemagno/InAppBillingPlugin/issues/512

It seems he is not going to add Avanolia support in that library.

Lunacy has in app purchase. 

Does anyone have any idea how they did it?

**Reply by grokys** [reply]
Regarding using `net6.0-mac` or `net6.0-maccatalyst` - I'm no expert here but I think you should be able to change the target framework of your Avalonia app to use one of these?

**Reply by robloo** [reply]
I looked at the Nuget, I bet you can use it with Avalonia for most platforms. As @grokys states, net6.0-mac is supported for example along with most other platforms. Have you tried pulling it into a non-Windows Avalonia platform and seeing if it works?

The only issue is native Windows support outside UWP/WinUI. However, I see no technical reason this can't be supported. Microsoft allows store API usage through desktop bridge apps.

**Reply by lanyusan** [reply]
> Regarding using `net6.0-mac` or `net6.0-maccatalyst` - I'm no expert here but I think you should be able to change the target framework of your Avalonia app to use one of these?

@ grokys

How do I do that on VS for Mac?

I tried to manually modify `csproj` file by setting 
`TargetFramework` to `net6.0-mac` or `net6.0-maccatalyst`, which couldn't be recognized by compiler.

```
  <PropertyGroup>
    <OutputType>WinExe</OutputType>
    <TargetFramework>net6.0-mac</TargetFramework>
    <Nullable>enable</Nullable>
    <!--Avalonia doesen't support TrimMode=link currently,but we are working on that https://github.com/AvaloniaUI/Avalonia/issues/6892 -->
    <TrimMode>copyused</TrimMode>
    <BuiltInComInteropSupport>true</BuiltInComInteropSupport>
  </PropertyGroup>
```

**Reply by lanyusan** [reply]
> I looked at the Nuget, I bet you can use it with Avalonia for most platforms. As @grokys states, net6.0-mac is supported for example along with most other platforms. Have you tried pulling it into a non-Windows Avalonia platform and seeing if it works?
> 
> The only issue is native Windows support outside UWP/WinUI. However, I see no technical reason this can't be supported. Microsoft allows store API usage through desktop bridge apps.

@robloo I have run the testing app on Mac, still the same issue `System.NotImplementedException `.

The testing app repo:

https://github.com/lanyusan/avanolia-iap-bug1

How do I change target framework to `net6.0-mac` or `net6.0-maccatalyst` ?

**Reply by maxkatz6** [reply]
> TargetFramework to net6.0-mac or net6.0-maccatalyst, which couldn't be recognized by compiler.

That was a correct way.
I have this sample project https://github.com/maxkatz6/avalonia-net6-macos-sample

**Reply by maxkatz6** [reply]
> The only issue is native Windows support outside UWP/WinUI. However, I see no technical reason this can't be supported. Microsoft allows store API usage through desktop bridge apps.

Yes, developers can use WinRT apis. The easiest way would be to target `net6.0-windows10.0.19041.0` (or another version) and use these APIs. Though, some API require app identity and packaging, which can be solved by creating a WAPPROJ project - which creates a MSIX package.

**Reply by maxkatz6** [reply]
BTW, avalonia doesn't work with maccatalyst atm. Instead, it supports "full APIs" - ios and macos.

**Comment by lanyusan** [comment]
@maxkatz6 

Greate news. I have tried your input on both Mac and Windows.

The plugin by James Montemagno can connect to app store on both platforms.

Thanks.

I haven't tested further the plugin by James but it looks now it would work.

**Reply by jamesmontemagno** [reply]
Feel free to make a PR in the readme on information on how to setup for Avalonia

**Comment by lanyusan** [comment]
After further investigation, it looks the plugin from James doesn't support `subscription` products for Windows app store.

https://github.com/jamesmontemagno/InAppBillingPlugin/issues/297

**Reply by robloo** [reply]
Again, please use the WinRT native API's directly. Microsoft also re-worked their store API's when they added subscriptions (was several years after the first version). You must be using Windows.Services.Store APIs instead of Windows.ApplicationModel.Store. I would guess James' plugin targets the old APIs.

**Reply by lanyusan** [reply]
Yes. I have decided to implement our own iap as we use subscription model.

**Reply by jamesmontemagno** [reply]
Yes, I just haven't had time to implement this into the plugin for Windows, however I am open to PRs for that.

**Reply by lanyusan** [reply]
> Yes, I just haven't had time to implement this into the plugin for Windows, however I am open to PRs for that.

I am pretty new to dotnet. I might be able to build an IAP function specifically for our needs but not sure if I can build it as a library. Would give it a try after we have got it work.

**Comment by renrutsirhc** [comment]
@lanyusan I'm also trying to get this working currently (subscription in app purchasing using windows.services.store api). I'm running into an issue which I've described in detail here: https://stackoverflow.com/questions/76096700/in-app-purchasing-with-microsoft-store-in-an-avalonia-app?noredirect=1#comment134206671_76096700 I'm wondering if you got your implementation working in the end?
